pcdsdevices/signal.py

```diff
@@ -428,23 +428,72 @@ class PVStateSignal(AggregateSignal):
 
     See `AggregateSignal` for more information.
     """
+    _metadata_keys = Signal._core_metadata_keys + (
+        'enum_strs',
+    )
 
     def __init__(self, *, name, **kwargs):
         super().__init__(name=name, **kwargs)
         for attr_name in self.parent._state_logic:
             self.add_signal_by_attr_name(attr_name)
+        self._has_setpoint_md = False
 
-    def describe(self):
+    @property
+    def enum_strs(self) -> Tuple[str, ...]:
+        """
+        Mimic the epics signal property for enum strings.
+        """
+        return tuple(state.name for state in self.parent.states_enum)
+
+    def describe(self) -> Dict[str, Dict[str, Any]]:
+        """
+        Make sure a reasonable description exists for bluesky scans.
+        """
         # Base description information
         sub_sigs = [sig.name for sig in self._signals]
-        desc = {'source': 'SUM:{}'.format(','.join(sub_sigs)),
-                'dtype': 'string',
-                'shape': [],
-                'enum_strs': tuple(state.name
-                                   for state in self.parent.states_enum)}
+        desc = {
+            'source': 'SUM:{}'.format(','.join(sub_sigs)),
+            'dtype': 'string',
+            'shape': [],
+            'enum_strs': self.enum_strs,
+        }
         return {self.name: desc}
 
-    def _calc_readback(self):
+    def _calc_readback(self) -> str:
+        """
+        Implements the PVStatePositioner logic.
+
+        This relies on this signal being a component of a PVStatePositioner
+        class with properly-defined class attributes.
+
+        On the first call, we'll do some setup of metadata values, since
+        this first call happens after the signals connect and we're guaranteed
+        that everything is ready. Doing it earlier can run into some
+        race conditions.
+        """
+        # Do some one-time setup here
+        # Convenient because we only hit this block when signals are ready
+        if (
+            self._metadata['enum_strs'] is None
+            and self.parent._state_initialized
+        ):
+            # One time setup of the enum strs
+            # Defined by class definition, not by EPICS
+            # Not necessarily available during init though, so do it now
+            self._metadata['enum_strs'] = self.enum_strs
+            self._run_metadata_callbacks()
+        if not self._has_setpoint_md:
+            # One time setup of the setpoint signal's metadata
+            # We need this to apply e.g. write permissions
+            ref_str = self.parent._state_logic_set_ref
+            if isinstance(ref_str, str):
+                setpoint_sig = getattr(self.parent, ref_str)
+                setpoint_sig.subscribe(
```

Comment:
Should stash the subscription identifier and unsubscribe on `destroy()`

Reply:
I'll make sure to do this

Reply:
Apparently `unsubscribe_all()` is called on all signals during `destroy()` whether or not this is handled here.
https://github.com/bluesky/ophyd/blob/0e8acb3df3d17e0ab7aa2cc924831f4a0c580449/ophyd/device.py#L1090-L1093
https://github.com/bluesky/ophyd/blob/0e8acb3df3d17e0ab7aa2cc924831f4a0c580449/ophyd/ophydobj.py#L345-L347